### PR TITLE
Fset 727: Reminder to complete the e-tray assessment

### PR DESCRIPTION
**Pull request by Lansalo**
Also include a partial refactoring around the service area for the reminder functionality as a whole

### CODE REVIEW

**Comment by Lansalo** [comment]
Still need to do some local testing after a merge into master and conflict resolution

File: app/connectors/EmailClient.scala [hunk 1]
```diff
@@ -178,11 +163,28 @@ trait CSREmailClient extends OnlineTestEmailClient with AssessmentCentreEmailCli
...
+  self: EmailClient =>
+
   def sendOnlineTestInvitation(to: String, name: String, expireDateTime: DateTime)(implicit hc: HeaderCarrier): Future[Unit]
   def sendOnlineTestExpired(to: String, name: String)(implicit hc: HeaderCarrier): Future[Unit]
-  def sendTestExpiringReminder(to: String, name: String, timeLeft: Int,
-    timeUnit: TimeUnit, expiryDate: DateTime)(implicit hc: HeaderCarrier): Future[Unit]
+  def sendTestExpiringReminder(to: String, name: String, timeLeftInHours: Int,
+                               timeUnit: TimeUnit, expiryDate: DateTime)(implicit hc: HeaderCarrier): Future[Unit]
+
+  protected def sendExpiringReminder(template: String)(to: String, name: String, timeLeftInHours: Int,
```
Comment (lukastymo) [inline app/connectors/EmailClient.scala]: Why not include template in the rest of parameters? Is Currying helping in something here?

File: app/model/persisted/NotificationExpiringOnlineTest.scala [hunk 2]
```diff
@@ -35,8 +36,8 @@ object NotificationExpiringOnlineTest {
     val personalDetailsRoot = doc.getAs[BSONDocument]("personal-details").get
     val preferredName = personalDetailsRoot.getAs[String]("preferredName").get
     val testGroupsRoot = doc.getAs[BSONDocument]("testGroups").get
-    val PHASE1Root = testGroupsRoot.getAs[BSONDocument]("PHASE1").get
-    val expiryDate = PHASE1Root.getAs[DateTime]("expirationDate").get
+    val PHASERoot = testGroupsRoot.getAs[BSONDocument]("PHASE1").getOrElse(testGroupsRoot.getAs[BSONDocument]("PHASE2").get)
```
Comment (lukastymo) [inline app/model/persisted/NotificationExpiringOnlineTest.scala]: We will never get Phase2 if Phase1 is present. It doesn't seem right.

File: app/model/ReminderNotice.scala [hunk 3]
```diff
@@ -16,20 +16,35 @@
...
+  private def timeUnitAndPhase: (TimeUnit, String) = progressStatuses match {
+    case PHASE1_TESTS_SECOND_REMINDER => (HOURS, phase_1)
+    case PHASE1_TESTS_FIRST_REMINDER => (DAYS, phase_1)
+    case PHASE2_TESTS_SECOND_REMINDER => (HOURS, phase_2)
+    case PHASE2_TESTS_FIRST_REMINDER => (DAYS, phase_2)
   }
+
+  val timeUnit: TimeUnit = timeUnitAndPhase._1
+  val phase: String = timeUnitAndPhase._2
+  val isPhase1 = phase == phase_1
```
Comment (lukastymo) [inline app/model/ReminderNotice.scala]: val isPhase1, val isPhase2 unused

File: app/model/ReminderNotice.scala [hunk 4]
```diff
@@ -16,20 +16,35 @@
 
 package model
 
-import model.ProgressStatuses.{ PHASE1_TESTS_FIRST_REMINDER, PHASE1_TESTS_SECOND_REMINDER, ProgressStatus }
+import model.ProgressStatuses._
 
 import scala.concurrent.duration.{ DAYS, HOURS, TimeUnit }
 
 sealed case class ReminderNotice(hoursBeforeReminder: Int, progressStatuses: ProgressStatus) {
```
Comment (lukastymo) [inline app/model/ReminderNotice.scala]: just small observation, don't need to change it:)

with sealed trait you won't need pattern matching and required. Each of the object would deliver exactly the thing the reminder needs

File: app/services/onlinetesting/Phase1TestService.scala [hunk 5]
```diff
@@ -74,6 +70,23 @@ trait Phase1TestService extends OnlineTestService with ResetPhase1Test {
...
+      case None => Future.successful(())
+    }
+  }
+
+  override def emailCandidateForExpiringTestReminder(expiringTest: NotificationExpiringOnlineTest,
+                                                     emailAddress: String,
+                                                     reminder: ReminderNotice)(implicit hc: HeaderCarrier): Future[Unit] = {
+    emailClient.sendTestExpiringReminder(emailAddress, expiringTest.preferredName,
+      reminder.hoursBeforeReminder, reminder.timeUnit, expiringTest.expiryDate).map { _ =>
+      audit(s"ReminderPhase1ExpiringOnlineTestNotificationBefore${reminder.hoursBeforeReminder}HoursEmailed",
```
Comment (lukastymo) [inline app/services/onlinetesting/Phase1TestService.scala]: There is non functional requirement to emit the DataStoreEvent everything something is modified. All non HTTP GET calls. To make this changeset smaller, you can do this in a separate commit.

File: app/model/ReminderNotice.scala (hunk 3 above)
Comment (Lansalo) [inline app/model/ReminderNotice.scala]: Done

File: app/model/ReminderNotice.scala [hunk 6]
```diff
@@ -16,20 +16,35 @@
...
+  require(progressStatuses == PHASE1_TESTS_FIRST_REMINDER ||
+          progressStatuses == PHASE1_TESTS_SECOND_REMINDER ||
+          progressStatuses == PHASE2_TESTS_FIRST_REMINDER ||
+          progressStatuses == PHASE2_TESTS_SECOND_REMINDER,
+          "progressStatuses value not allowed")
 
-  def timeUnit: TimeUnit = progressStatuses match {
-    case PHASE1_TESTS_SECOND_REMINDER => HOURS
-    case PHASE1_TESTS_FIRST_REMINDER => DAYS
+  private val phase_1 = "PHASE1"
```
Comment (Lansalo) [inline app/model/ReminderNotice.scala]: Done

File: conf/application.conf [hunk 7]
```diff
@@ -193,6 +193,18 @@ scheduling {
       initialDelaySecs = 30
       intervalSecs = 30
     }
+    first-phase2-reminder-expiring-test-job {
```
Comment (lukastymo) [inline conf/application.conf]: Remember to add also app-config-dev/qa/staging/prod for these two

File: app/connectors/EmailClient.scala (hunk 1 above)
Comment (Lansalo) [inline app/connectors/EmailClient.scala]: The advantage here that the phase1/pase2 client vill populate the template value so the caller won't need to bother about it.

File: conf/application.conf [hunk 8]
```diff
@@ -193,6 +193,18 @@ scheduling {
...
+      enabled = false
+      lockId = "first-phase2-reminder-lock-coordinator"
+      initialDelaySecs = 31
```
Comment (lukastymo) [inline conf/application.conf]: Any reason why it's 31 not 30 like all others?

File: app/model/persisted/NotificationExpiringOnlineTest.scala (hunk 2 above)
Comment (Lansalo) [inline app/model/persisted/NotificationExpiringOnlineTest.scala]: Done

File: conf/application.conf (hunk 8 above)
Comment (Lansalo) [inline conf/application.conf]: No... just thought not to have all of them starting at the same time. Shouldn't be a problem as long as the interval is the same.
